### PR TITLE
ESM + CJS envs APIs

### PR DESCRIPTION
**Pull request by GiladShoham**
## Proposed Changes
This PR is laying the infrastructure and API in order to move the core browser-related envs (react, HTML, mdx) to ESM by default.
This PR doesn't change the current behavior of the envs, it just provides various API around both esm and cjs, and prepare different configs, so it will be easier to change the default in the future.

These APIs provide an easy way to build (right now) a custom env, which is ESM or CJS.
This serves a few purposes:
1. Allow users, to build their own cjs envs with the new API, so in the future when we will change the default, they won't be suddenly affected by it.
2. Allow users to build their own ESM envs, so they can:
  a. enjoy ESM right away (without the need to wait for us), ESM has several benefits like better tree shaking and code splitting.
  b. test how ESM affect them today, so they can make all the required changes and be ready for our change.    

In this PR we also changed the node/aspect envs to use the new APIs so they won't be affected when changing the default.

Examples of how to use the new APIs will be added to the docs soon.

#### Typescript
- createCjsCompiler - Create a compiler instance and run the cjs transformer for it
- createEsmCompiler - Create a compiler instance and run the esm transformer for it
- getCjsTransformer - Create a transformer that change the ts module to CommonJS
- getEsmTransformer - Create a transformer that change the ts module to ES2020
- getCjsPackageJsonProps - add the default package json properties to the component
- getEsmPackageJsonProps - add type: module to the package.json props and the default props

#### React
- getCjsJestTester - Get a jest tester instance with react config and cjs configs
- getEsmJestTester - Get a jest tester instance with react config and esm configs
- getTsCjsCompiler - Get a compiler instance with react config and set it to cjs module
- getTsEsmCompiler - Get a compiler instance with react config and set it to esm module
- getCjsPackageJsonProps - Get the default package.json props for a cjs component
- getEsmPackageJsonProps - Get the default package.json props for an esm component
- getBuildPipeWithoutCompiler - Get the react build pipeline without the compilation task.
- getCjsCompilerTask - Get a compiler task with react config and set to cjs module
- getEsmCompilerTask - Get a compiler task with react config and set to esm module
- re-organize all jest configs and provide both esm and cjs configs separately 

#### Compiler
- add getDistDir API for compiler interface
- add getRelativeDistFolder - find the compiler configured on the workspace and ask for the dist folder path
- add isDistDirExists - Check if the dist folder (in the component package under node_modules) exist

#### Pkg
- isModulePathExists - check if the module for the component exist in node_modules

#### Typescript config mutator
- add APIs to tsconfig mutator - setModule, setOutDir. expose ModuleKind enum

#### Envs
- getOrCalculateEnv - This will try to use the regular getEnv but fallback to the calculate env (in case you are using it during on load)

#### Typescript config mutator
- change setDistDir API for tsconfig mutator to apply on all relevant props

### Dependencies upgrades
- upgrade all jest related packages

### Other changes
- make node env to use cjs by utilizing new react APIs (make sure we won't break it when changing react default in the future)
- make aspect env to use cjs by utilizing new react APIs (make sure we won't break it when changing react default in the future)
- use compiler dist dir when compiling components in workspace
- use the new ts aspect apis to get esm/cjs compiler from react env
- fix preview link files to use the components' compilers dist folder
- fix react override package json props func
- move calculate components issues - missing dists and missing links to compiler and pkg aspects

### CODE REVIEW

File: scopes/harmony/node/node.env.ts [hunk 1]
```diff
@@ -21,4 +24,8 @@ export class NodeEnv implements DependenciesEnv {
       splitComponentBundle: false,
     };
   }
+
+  getPackageJsonProps() {
+    return this.tsAspect.getCjsPackageJsonProps();
```
Comment (KutnerUri) [inline scopes/harmony/node/node.env.ts:29]: in aspect env you accomplish this with
```
this.reactEnv.getCjsPackageJsonProps();
```

let's use the same code, if possible

File: scopes/typescript/typescript/typescript.main.runtime.ts [hunk 2]
```diff
@@ -114,9 +140,18 @@ export class TypescriptMain {
...
+      main: 'dist/{main}.js',
+      types: '{main}.ts',
+    };
+  }
+
+  getEsmPackageJsonProps(): PackageJsonProps {
     return {
+      // main: 'dist-esm/{main}.js',
       main: 'dist/{main}.js',
+      type: 'module',
```
Comment (KutnerUri) [inline scopes/typescript/typescript/typescript.main.runtime.ts:154]: we should be careful with this property, it causes errors, usually when packaging both cjs and mjs in the same package.
I think it's ok if the package is fully esm.

maybe we should set ```type: 'commonjs'``` for cjs? reference - https://nodejs.org/docs/latest-v13.x/api/esm.html#esm_package_json_type_field

File: scopes/react/react/jest/transformer.js [hunk 3]
```diff
@@ -17,7 +18,6 @@ const presets = [
   ],
 ];
 const plugins = [
-  [require('@babel/plugin-transform-modules-commonjs')],
```
Comment (KutnerUri) [inline scopes/react/react/jest/transformer.js:20]: 👍

File: scopes/compilation/compiler/types.ts [hunk 4]
```diff
@@ -96,6 +96,11 @@ export interface Compiler extends CompilerOptions {
    */
   build(context: BuildContext): Promise<BuiltTaskResult>;
 
+  /**
+   * return the dist dir of the compiled files
+   */
+  getDistDir(): string;
```
Comment (KutnerUri) [inline scopes/compilation/compiler/types.ts:102]: please document - abs or relative path? and relative to what?

File: scopes/envs/envs/environments.main.runtime.ts [hunk 5]
```diff
@@ -218,6 +218,19 @@ export class EnvsMain {
     throw new EnvNotFound(id, component.id.toString());
   }
 
+  /**
+   * get the env of the given component.
+   * This will try to use the regular getEnv but fallback to the calculate env (in case you are using it during on load)
+   * This is safe to be used on onLoad as well
+   */
+  getOrCalculateEnv(component: Component): EnvDefinition {
```
Comment (KutnerUri) [inline scopes/envs/envs/environments.main.runtime.ts:226]: can't we add this functionality to the existing getEnv?

File: scopes/typescript/modules/ts-config-mutator/ts-config-mutator.ts [hunk 6]
```diff
@@ -80,6 +71,31 @@ export class TypescriptConfigMutator {
     return this;
   }
 
+  setModule(module: string): TypescriptConfigMutator {
```
Comment (KutnerUri) [inline scopes/typescript/modules/ts-config-mutator/ts-config-mutator.ts:74]: jsdocs please - what is this meant for?

File: scopes/typescript/typescript/typescript.main.runtime.ts [hunk 7]
```diff
@@ -103,6 +103,32 @@ export class TypescriptMain {
...
+    return this.createCompiler(options, [this.getCjsTransformer(), ...transformers], tsModule);
+  }
+
+  createEsmCompiler(options: TypeScriptCompilerOptions, transformers: TsConfigTransformer[] = [], tsModule = ts) {
+    return this.createCompiler(options, [this.getEsmTransformer(), ...transformers], tsModule);
+  }
+
+  getCjsTransformer(): TsConfigTransformer {
+    const cjsTransformer = (config: TypescriptConfigMutator) => {
+      // config.setDistDir('dist');
```
Comment (KutnerUri) [inline scopes/typescript/typescript/typescript.main.runtime.ts:116]: cleanup

File: scopes/workspace/workspace/workspace-component/workspace-component-loader.ts [hunk 8]
```diff
@@ -272,7 +272,7 @@ export class WorkspaceComponentLoader {
 
   private async upsertExtensionData(component: Component, extension: string, data: any) {
     const existingExtension = component.state.config.extensions.findExtension(extension);
-    if (existingExtension) {
+    if (existingExtension && data) {
```
Comment (KutnerUri) [inline scopes/workspace/workspace/workspace-component/workspace-component-loader.ts:275]: didn't this allow removing data by setting undefined?

**Comment by KutnerUri** [comment]
please 
- [ ] show an example of composing react env and setting as ESM/CJS (also in docs?)

File: scopes/typescript/typescript/typescript.main.runtime.ts (hunk 2 above)
Comment (GiladShoham) [reply scopes/typescript/typescript/typescript.main.runtime.ts:154]: That's a dedicated API to do exactly this.
it called `getEsmPackageJsonProps` for a reason.. 
it should be used by an env only if you compile to esm..

File: scopes/harmony/node/node.env.ts (hunk 1 above)
Comment (GiladShoham) [reply scopes/harmony/node/node.env.ts:60]: done

File: scopes/envs/envs/environments.main.runtime.ts (hunk 5 above)
Comment (GiladShoham) [reply scopes/envs/envs/environments.main.runtime.ts:226]: not exactly, there is a reason for this.
Anyway it's not part of this PR

File: scopes/workspace/workspace/workspace-component/workspace-component-loader.ts (hunk 8 above)
Comment (GiladShoham) [reply scopes/workspace/workspace/workspace-component/workspace-component-loader.ts:275]: no. you should never use this to remove data. by design

File: scopes/typescript/typescript/typescript.main.runtime.ts (hunk 7 above)
Comment (GiladShoham) [reply scopes/typescript/typescript/typescript.main.runtime.ts]: done

File: scopes/typescript/modules/ts-config-mutator/ts-config-mutator.ts (hunk 6 above)
Comment (GiladShoham) [reply scopes/typescript/modules/ts-config-mutator/ts-config-mutator.ts:80]: done

File: scopes/compilation/compiler/types.ts (hunk 4 above)
Comment (GiladShoham) [reply scopes/compilation/compiler/types.ts]: done

**Comment by GiladShoham** [comment]
> please
> 
> * [ ]  show an example of composing react env and setting as ESM/CJS (also in docs?)

will do

**Pull request merged**